python/task_manager/worker_thread.py

```diff
@@ -95,18 +97,19 @@ def run(self):
                     if not self._process_tasks:
                         break
                     # emit the result (non-blocking):
-                    self._results_poller.queue_task_completed(self, task_to_process, result)
+                    self._results_dispatcher.produce_event(TaskCompletedEvent(self, task_to_process, result, self._results_dispatcher))
```

Comment:
Hmm, having looked at the code, it feels to me that we may be making this a lot more complicated just because we are introducing these `TaskCompletedEvent` and `TaskFailedEvent` classes. Do we need them? Can they just be internal to the results dispatcher?

Maybe doing something like this is simpler:

```
self._results_dispatcher.signal_completion(task_to_process, result))
self._results_dispatcher.signal_failure(task_to_process, str(e), tb))
```

This seems conceptually a little clearer to me - maybe we can discuss?